File: matlab/preprocessing/identify_brake_sensors.m
function [Nodo_out, rolesTable, steadyInfo] = identify_brake_sensors(Nodo, varargin)
% IDENTIFY_BRAKE_SENSORS — steady-window classifier with hard constraints
% Spec:
%   • LPF 1 Hz (Butter 4th, zero-phase) on a uniform 5 Hz grid
%   • Compute dP/dt (bar/s)
%   • Find contiguous "steady" regions where for ALL sensors:
%       |dP/dt| < 0.01  AND nearest raw sample ≤ 3 s
%   • Apply only on a steady region that contains any pressure > 4.6 bar
%   • Window can be short; if longer than 10 min, clip to 10 min around max pressure
%   • Classify by median pressure in that window:
%       > 4.6 → MBP,  < 0.2 → BC,  else → WV
%   • Enforce: exactly 1 MBP; equal numbers of WV and BC among the rest
%   • OUTPUT: same Nodo struct but with .Label, sorted: MBP first, WV next, BC last.
%   • If no candidate steady region is found, **skip** classification and
%     return original order with empty labels and an informative message.

% ---- parameters ----
p = inputParser;
addParameter(p,'FsGrid',5);
addParameter(p,'Fc',1);
addParameter(p,'GradTol',0.01);
addParameter(p,'MaxGapSec',3);
addParameter(p,'MaxWindowMin',10);
addParameter(p,'MinWindowSec',15);  % prefer blocks >= this (fallback allowed)
parse(p,varargin{:});
Fs       = p.Results.FsGrid;
Fc       = p.Results.Fc;
GradTol  = p.Results.GradTol;
MaxGap   = p.Results.MaxGapSec;
MaxWin   = p.Results.MaxWindowMin;
MinWin   = p.Results.MinWindowSec;

n = numel(Nodo);
if n==0, error('Empty Nodo.'); end

% ---- thin arrays & overlap window ----
ID = strings(n,1); T = cell(n,1); P = cell(n,1);
for k = 1:n
    ID(k) = string(Nodo(k).ID);
    t = Nodo(k).Time;       if isrow(t), t = t.'; end
    x = Nodo(k).Pressure;   if isrow(x), x = x.'; end
    good = ~isnat(t) & isfinite(x);
    t = t(good); x = x(good);

    % >>> NEW: enforce strictly increasing, unique time stamps
    if ~isempty(t)
        [t, ord] = sort(t, 'ascend');    % sort by time
        x = x(ord);
        [t, ia] = unique(t, 'stable');   % drop duplicates
        x = x(ia);
    end

    if numel(t) < 2
        % not enough points to interpolate linearly → graceful skip
        [Nodo_out, rolesTable, steadyInfo] = skip_and_fill_empty(Nodo, ID, ...
            sprintf('Sensor %s has <2 unique time samples.', ID(k)));
        return
    end

    T{k} = t; P{k} = x;
end

t0 = max(cellfun(@(t)t(1), T));
t1 = min(cellfun(@(t)t(end), T));
if isnat(t0) || isnat(t1) || t0>=t1
    [Nodo_out, rolesTable, steadyInfo] = skip_and_fill_empty(Nodo, ID, 'No temporal overlap across sensors.');
    return
end

% ---- grid & filter ----
dt = seconds(1/Fs);
tgrid = (t0:dt:t1).';  m = numel(tgrid);
Wn = Fc/(Fs/2);
if Wn>=1
    [Nodo_out, rolesTable, steadyInfo] = skip_and_fill_empty(Nodo, ID, 'Invalid LPF setup: Fc must be < Fs/2.');
    return
end
[b,a] = butter(4, Wn);

Pr = nan(m,n); Pf = nan(m,n); dP = nan(m,n); validNear = false(m,n);
t0s = seconds(tgrid - tgrid(1));

for k = 1:n
    % >>> NEW: re-assert sorted/unique before building timetable (extra safety)
    tk = T{k}; pk = P{k};
    [tk, ord] = sort(tk, 'ascend'); pk = pk(ord);
    [tk, ia]  = unique(tk, 'stable'); pk = pk(ia);

    % retime to grid (requires strictly monotonic times)
    TT  = timetable(tk, pk, 'VariableNames', {'P'});
    TT1 = retime(TT, tgrid, 'linear');
    y   = TT1.P; Pr(:,k) = y;

    % nearest-sample check ≤ MaxGap
    tkS = seconds(tk - tgrid(1));
    idxNear = round(interp1(tkS, (1:numel(tkS))', t0s, 'nearest', 'extrap'));
    idxNear = max(1, min(idxNear, numel(tkS)));
    validNear(:,k) = abs(t0s - tkS(idxNear)) <= MaxGap;

    % simple inpaint for filtering only (keep NaN after)
    yi = y;
    nanMask = isnan(yi);
    if any(nanMask)
        ii = find(~nanMask);
        if ~isempty(ii)
            yi(nanMask) = interp1(ii, yi(ii), find(nanMask), 'linear', 'extrap');
        end
    end

    if all(~isnan(yi))
        yf = filtfilt(b,a, yi);
    else
        yf = yi;
    end

    % restore NaN where not supported by raw neighborhood
    yf(~validNear(:,k)) = NaN;
    Pf(:,k) = yf;

    % gradient (bar/s)
    g = [0; diff(yf)] * Fs;
    bad = isnan(yf) | [false; isnan(yf(1:end-1))];
    g(bad) = NaN;
    dP(:,k) = g;
end

% ---- steady mask (all sensors must be steady & valid) ----
steadyPerSensor = (abs(dP) < GradTol) & validNear;
steadyAll = all(steadyPerSensor, 2);

% ---- find steady blocks that contain any P>4.6 ----
blocks = contiguous_blocks(steadyAll);
cand = [];
for b = 1:size(blocks,1)
    i1b = blocks(b,1); i2b = blocks(b,2);
    if seconds(tgrid(i2b)-tgrid(i1b)) < MinWin, continue; end
    seg = Pf(i1b:i2b, :);
    if any(seg(:) > 4.6)
        cand = [cand; i1b i2b]; %#ok<AGROW>
    end
end
if isempty(cand)
    % fallback: allow even shorter steady blocks if they contain P>4.6
    for b = 1:size(blocks,1)
        i1b = blocks(b,1); i2b = blocks(b,2);
        seg = Pf(i1b:i2b, :);
        if any(seg(:) > 4.6)
            cand = [cand; i1b i2b]; %#ok<AGROW>
        end
    end
end
if isempty(cand)
    % Graceful skip: no candidate steady window with P>4.6
    [Nodo_out, rolesTable, steadyInfo] = skip_and_fill_empty(Nodo, ID, ...
        'No steady window found that contains any pressure > 4.6 bar.');
    % Add basic context to steadyInfo
    steadyInfo.tgrid      = tgrid;
    steadyInfo.P_filtered = Pf;
    steadyInfo.dP_dt      = dP;
    steadyInfo.steadyAll  = steadyAll;
    return
end

% pick candidate with the longest duration
dur = cand(:,2) - cand(:,1) + 1;
[~, j] = max(dur);
i1 = cand(j,1); i2 = cand(j,2);

% clip to MaxWindowMin around time of maximum pressure within block
maxWinSamples = round(MaxWin * 60 * Fs);
segP = Pf(i1:i2, :);
[~, idxLocalMax] = max(max(segP, [], 2));
center = i1 + idxLocalMax - 1;
if (i2 - i1 + 1) > maxWinSamples
    half = floor(maxWinSamples/2);
    a = max(center - half, i1);
    b = min(a + maxWinSamples - 1, i2);
    i1 = a; i2 = b;
end
steadyMask = false(m,1); steadyMask(i1:i2) = true;

% ---- classification by median pressure in chosen steady window ----
MedianP = nan(n,1);
for k = 1:n
    pk = Pf(steadyMask, k);
    pk = pk(~isnan(pk));
    if isempty(pk), MedianP(k) = NaN; else, MedianP(k) = median(pk); end
end

roles = repmat("WV", n, 1);          % initial
roles(MedianP > 4.6) = "MBP";
roles(MedianP < 0.2) = "BC";
roles(~isfinite(MedianP)) = "Unknown";

% ---- enforce exactly ONE MBP ----
MBP_idx = find(roles=="MBP");
if isempty(MBP_idx)
    [~, imax] = max(MedianP);
    if isfinite(MedianP(imax))
        roles(imax) = "MBP";
        MBP_idx = imax;
    end
elseif numel(MBP_idx) > 1
    [~, jmax] = max(MedianP(MBP_idx));
    keep = MBP_idx(jmax);
    flip = setdiff(MBP_idx, keep);
    roles(flip) = "WV";
    MBP_idx = keep;
end

% ---- enforce equal BC and WV among the rest ----
rest = setdiff(1:n, MBP_idx);
isBC = roles(rest)=="BC";  isWV = roles(rest)=="WV";
numRest = numel(rest);
target = floor(numRest/2);  % target per class

% too many BC → move those closest to 0.2 up to WV
excessBC = sum(isBC) - target;
if excessBC > 0
    rc = rest(isBC);
    [~,ord] = sort(MedianP(rc),'descend','MissingPlacement','last');
    move = rc(ord(1:excessBC));
    roles(move) = "WV";
end
% too few BC → move WV with smallest MedianP down to BC
isBC = roles(rest)=="BC";
needBC = target - sum(isBC);
if needBC > 0
    rc = rest(roles(rest)=="WV");
    [~,ord] = sort(MedianP(rc),'ascend','MissingPlacement','last');
    move = rc(ord(1:needBC));
    roles(move) = "BC";
end

% ---- produce outputs: LABEL + SORTED Nodo ----
labelStr = cellstr(roles);  % cell array of char
% attach .Label to each entry
NodoL = Nodo;
for k = 1:n
    NodoL(k).Label = labelStr{k};
end

% sorting order: MBP first, WV next, BC last (Unknown, if any, appended at end)
MBP_idx = find(roles=="MBP");
WV_idx  = find(roles=="WV");
BC_idx  = find(roles=="BC");
UNK_idx = find(roles=="Unknown");

order = [MBP_idx(:); WV_idx(:); BC_idx(:); UNK_idx(:)];
Nodo_out = NodoL(order);

% ---- roles table (after sorting) ----
rolesTable = table(order, string({Nodo(order).ID})', string(labelStr(order)), MedianP(order), ...
    'VariableNames', {'OriginalIdx','ID','Label','MedianP_Steady'});

% ---- command-window output ----
fprintf('\n[identify_brake_sensors]\n');
fprintf('  Steady window used: %s  →  %s  (%.1f min)\n', ...
    datestr(tgrid(i1)), datestr(tgrid(i2)), seconds(tgrid(i2)-tgrid(i1))/60);
fprintf('  Grid Fs = %.1f Hz, LPF Fc = %.1f Hz, |dP/dt| < %.3f bar/s, MaxGap ≤ %.0f s\n', ...
    Fs, Fc, GradTol, MaxGap);
fprintf('  Classification by median pressure in steady window: >4.6=MBP, <0.2=BC, else=WV\n');
disp(rolesTable);

% ---- debug / plotting info ----
steadyInfo = struct();
steadyInfo.tgrid       = tgrid;
steadyInfo.P_filtered  = Pf;
steadyInfo.dP_dt       = dP;
steadyInfo.steadyAll   = steadyAll;
steadyInfo.steadyMask  = steadyMask;
steadyInfo.window_idx  = [i1 i2];
steadyInfo.window_time = [tgrid(i1) tgrid(i2)];
steadyInfo.MaxWindowMin= MaxWin;
steadyInfo.candidateFound = true;

end

% -------------------- helpers --------------------
function arr = contiguous_blocks(x)
% return [iStart iEnd] rows for each contiguous run of true in logical x
    x = x(:);
    if ~any(x), arr = zeros(0,2); return; end
    d = diff([false; x; false]);
    s = find(d==1);
    e = find(d==-1)-1;
    arr = [s e];
end

function [Nodo_out, rolesTable, steadyInfo] = skip_and_fill_empty(Nodo, ID, reason)
% Return original Nodo with empty .Label, empty-ish rolesTable, and message.
    n = numel(Nodo);
    Nodo_out = Nodo;
    for k = 1:n
        Nodo_out(k).Label = '';   % empty label as requested
    end
    rolesTable = table((1:n).', ID, repmat(string(''), n,1), repmat(NaN,n,1), ...
        'VariableNames', {'OriginalIdx','ID','Label','MedianP_Steady'});

    fprintf('\n[identify_brake_sensors] SKIPPED: %s\n', reason);
    if n>0
        t0 = Nodo(1).Time(1); t1 = Nodo(1).Time(end);
        if ~isnat(t0) && ~isnat(t1)
            fprintf('  Data span (sensor 1): %s → %s (%.1f h)\n', ...
                datestr(t0), datestr(t1), hours(t1 - t0));
        end
    end
    disp(rolesTable);

    steadyInfo = struct();
    steadyInfo.candidateFound = false;
    steadyInfo.reason = reason;
end

File: python_port/ingestion/identify_brake_sensors.py
"""Python port of matlab/preprocessing/identify_brake_sensors.m.

Steady-window classifier with hard constraints:
  - LPF 1 Hz (Butterworth 4th order, zero-phase) on a uniform 5 Hz grid
  - Compute dP/dt (bar/s)
  - Find contiguous "steady" regions where for ALL sensors:
      |dP/dt| < 0.01  AND nearest raw sample <= 3 s away
  - Apply only on a steady region that contains any pressure > 4.6 bar
  - Window can be short; if longer than 10 min, clip to 10 min around max pressure
  - Classify by median pressure in that window: >4.6 -> MBP, <0.2 -> BC, else -> WV
  - Enforce: exactly 1 MBP; equal numbers of WV and BC among the rest
  - If no candidate steady region is found, skip classification and return the
    original order with empty labels.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, List, Optional, Tuple

import numpy as np
from scipy.signal import butter, filtfilt


@dataclass
class SteadyInfo:
    candidate_found: bool
    reason: Optional[str] = None
    tgrid: Optional[np.ndarray] = None
    p_filtered: Optional[np.ndarray] = None
    dp_dt: Optional[np.ndarray] = None
    steady_all: Optional[np.ndarray] = None
    steady_mask: Optional[np.ndarray] = None
    window_idx: Optional[Tuple[int, int]] = None
    window_time: Optional[Tuple[Any, Any]] = None
    max_window_min: Optional[float] = None
    from_cache: bool = False


def _contiguous_blocks(x: np.ndarray) -> np.ndarray:
    """Return Nx2 array of [start, end] (inclusive, 0-based) for runs of True in x."""
    x = np.asarray(x, dtype=bool)
    if not x.any():
        return np.zeros((0, 2), dtype=int)
    d = np.diff(np.concatenate(([False], x, [False])).astype(int))
    starts = np.where(d == 1)[0]
    ends = np.where(d == -1)[0] - 1
    return np.column_stack([starts, ends])


def _nearest_gap_seconds(tk_seconds: np.ndarray, query_seconds: np.ndarray) -> np.ndarray:
    """abs(query - nearest tk) for each query point, clamped at the array bounds
    (mirrors interp1(tkS, idx, t0s, 'nearest', 'extrap'))."""
    n = len(tk_seconds)
    idx = np.searchsorted(tk_seconds, query_seconds)
    idx_hi = np.clip(idx, 0, n - 1)
    idx_lo = np.clip(idx - 1, 0, n - 1)
    d_hi = np.abs(query_seconds - tk_seconds[idx_hi])
    d_lo = np.abs(query_seconds - tk_seconds[idx_lo])
    return np.minimum(d_hi, d_lo)


def _retime_linear(tk_seconds: np.ndarray, pk: np.ndarray, tgrid_seconds: np.ndarray) -> np.ndarray:
    """Linear interpolation onto tgrid, NaN outside [tk_seconds[0], tk_seconds[-1]]
    (mirrors MATLAB timetable retime(..., 'linear'), which does not extrapolate)."""
    return np.interp(tgrid_seconds, tk_seconds, pk, left=np.nan, right=np.nan)


def sort_by_label(nodo: List[dict]) -> List[dict]:
    """Sort MBP first, then WV, then BC (Unknown/blank last)."""
    order_rank = {"MBP": 0, "WV": 1, "BC": 2}
    return sorted(nodo, key=lambda s: order_rank.get(str(s.get("Label") or ""), 3))


def _skip_and_fill_empty(nodo: List[dict], ids: List[str], reason: str) -> Tuple[List[dict], List[dict], SteadyInfo]:
    nodo_out = [dict(s) for s in nodo]
    for s in nodo_out:
        s["Label"] = ""
    roles_table = [
        {"OriginalIdx": i, "ID": ids[i], "Label": "", "MedianP_Steady": np.nan}
        for i in range(len(nodo))
    ]
    print(f"\n[identify_brake_sensors] SKIPPED: {reason}")
    if nodo and nodo[0].get("Time") is not None and len(nodo[0]["Time"]) > 0:
        t0, t1 = nodo[0]["Time"][0], nodo[0]["Time"][-1]
        if not (np.isnat(t0) or np.isnat(t1)):
            span_h = (t1 - t0) / np.timedelta64(1, "h")
            print(f"  Data span (sensor 1): {t0} -> {t1} ({span_h:.1f} h)")
    return nodo_out, roles_table, SteadyInfo(candidate_found=False, reason=reason)


def identify_brake_sensors(
    nodo: List[dict],
    fs_grid: float = 5.0,
    fc: float = 1.0,
    grad_tol: float = 0.01,
    max_gap_sec: float = 3.0,
    max_window_min: float = 10.0,
    min_window_sec: float = 15.0,
) -> Tuple[List[dict], List[dict], SteadyInfo]:
    """Returns (nodo_out, roles_table, steady_info). nodo_out entries get a
    'Label' field and are re-ordered MBP -> WV -> BC (-> Unknown)."""
    n = len(nodo)
    if n == 0:
        raise ValueError("Empty Nodo.")

    ids = [str(s["ID"]) for s in nodo]
    T: List[np.ndarray] = [None] * n
    P: List[np.ndarray] = [None] * n

    for k in range(n):
        t = np.asarray(nodo[k]["Time"]).ravel()
        x = np.asarray(nodo[k]["Pressure"], dtype=float).ravel()
        good = ~np.isnat(t) & np.isfinite(x)
        t, x = t[good], x[good]

        if len(t):
            order = np.argsort(t, kind="stable")
            t, x = t[order], x[order]
            t, unique_idx = np.unique(t, return_index=True)
            x = x[unique_idx]

        if len(t) < 2:
            return _skip_and_fill_empty(nodo, ids, f"Sensor {ids[k]} has <2 unique time samples.")

        T[k], P[k] = t, x

    t0 = max(t[0] for t in T)
    t1 = min(t[-1] for t in T)
    if np.isnat(t0) or np.isnat(t1) or t0 >= t1:
        return _skip_and_fill_empty(nodo, ids, "No temporal overlap across sensors.")

    # ---- grid & filter ----
    dt = np.timedelta64(int(round(1e9 / fs_grid)), "ns")
    tgrid = np.arange(t0, t1 + dt, dt)
    m = len(tgrid)
    wn = fc / (fs_grid / 2)
    if wn >= 1:
        return _skip_and_fill_empty(nodo, ids, "Invalid LPF setup: Fc must be < Fs/2.")
    b, a = butter(4, wn)

    tgrid_seconds = (tgrid - tgrid[0]) / np.timedelta64(1, "s")

    Pr = np.full((m, n), np.nan)
    Pf = np.full((m, n), np.nan)
    dP = np.full((m, n), np.nan)
    valid_near = np.zeros((m, n), dtype=bool)

    for k in range(n):
        tk_seconds = (T[k] - tgrid[0]) / np.timedelta64(1, "s")
        pk = P[k]

        y = _retime_linear(tk_seconds, pk, tgrid_seconds)
        Pr[:, k] = y

        gap = _nearest_gap_seconds(tk_seconds, tgrid_seconds)
        valid_near[:, k] = gap <= max_gap_sec

        # simple inpaint for filtering only (keep NaN after)
        yi = y.copy()
        nan_mask = np.isnan(yi)
        if nan_mask.any():
            good_idx = np.flatnonzero(~nan_mask)
            if good_idx.size:
                yi[nan_mask] = np.interp(np.flatnonzero(nan_mask), good_idx, yi[good_idx])

        if np.all(~np.isnan(yi)):
            yf = filtfilt(b, a, yi)
        else:
            yf = yi

        yf = yf.copy()
        yf[~valid_near[:, k]] = np.nan
        Pf[:, k] = yf

        g = np.concatenate(([0.0], np.diff(yf))) * fs_grid
        bad = np.isnan(yf) | np.concatenate(([False], np.isnan(yf[:-1])))
        g[bad] = np.nan
        dP[:, k] = g

    # ---- steady mask (all sensors must be steady & valid) ----
    steady_per_sensor = (np.abs(dP) < grad_tol) & valid_near
    steady_all = np.all(steady_per_sensor, axis=1)

    # ---- find steady blocks that contain any P>4.6 ----
    blocks = _contiguous_blocks(steady_all)
    min_win_td = np.timedelta64(int(round(min_window_sec * 1e9)), "ns")
    cand = []
    for i1b, i2b in blocks:
        if (tgrid[i2b] - tgrid[i1b]) < min_win_td:
            continue
        seg = Pf[i1b:i2b + 1, :]
        if np.any(seg > 4.6):
            cand.append((i1b, i2b))
    if not cand:
        # fallback: allow even shorter steady blocks if they contain P>4.6
        for i1b, i2b in blocks:
            seg = Pf[i1b:i2b + 1, :]
            if np.any(seg > 4.6):
                cand.append((i1b, i2b))
    if not cand:
        info = SteadyInfo(
            candidate_found=False,
            reason="No steady window found that contains any pressure > 4.6 bar.",
            tgrid=tgrid, p_filtered=Pf, dp_dt=dP, steady_all=steady_all,
        )
        nodo_out, roles_table, _ = _skip_and_fill_empty(
            nodo, ids, "No steady window found that contains any pressure > 4.6 bar."
        )
        return nodo_out, roles_table, info

    # pick candidate with the longest duration
    durations = [i2 - i1 + 1 for i1, i2 in cand]
    i1, i2 = cand[int(np.argmax(durations))]

    # clip to MaxWindowMin around time of maximum pressure within block
    max_win_samples = int(round(max_window_min * 60 * fs_grid))
    seg_p = Pf[i1:i2 + 1, :]
    idx_local_max = int(np.argmax(np.nanmax(seg_p, axis=1)))
    center = i1 + idx_local_max
    if (i2 - i1 + 1) > max_win_samples:
        half = max_win_samples // 2
        a_ = max(center - half, i1)
        b_ = min(a_ + max_win_samples - 1, i2)
        i1, i2 = a_, b_
    steady_mask = np.zeros(m, dtype=bool)
    steady_mask[i1:i2 + 1] = True

    # ---- classification by median pressure in chosen steady window ----
    median_p = np.full(n, np.nan)
    for k in range(n):
        pk = Pf[steady_mask, k]
        pk = pk[~np.isnan(pk)]
        if pk.size:
            median_p[k] = np.median(pk)

    roles = np.full(n, "WV", dtype=object)
    roles[median_p > 4.6] = "MBP"
    roles[median_p < 0.2] = "BC"
    roles[~np.isfinite(median_p)] = "Unknown"

    # ---- enforce exactly ONE MBP ----
    mbp_idx = np.flatnonzero(roles == "MBP")
    if mbp_idx.size == 0:
        imax = int(np.nanargmax(median_p)) if np.any(np.isfinite(median_p)) else None
        if imax is not None and np.isfinite(median_p[imax]):
            roles[imax] = "MBP"
            mbp_idx = np.array([imax])
    elif mbp_idx.size > 1:
        jmax = mbp_idx[np.argmax(median_p[mbp_idx])]
        flip = np.setdiff1d(mbp_idx, [jmax])
        roles[flip] = "WV"
        mbp_idx = np.array([jmax])

    # ---- enforce equal BC and WV among the rest ----
    rest = np.setdiff1d(np.arange(n), mbp_idx)
    num_rest = len(rest)
    target = num_rest // 2

    is_bc = roles[rest] == "BC"
    excess_bc = int(is_bc.sum()) - target
    if excess_bc > 0:
        rc = rest[is_bc]
        # sort descending by MedianP (NaN last), move the highest (closest to 0.2 from below)
        vals = median_p[rc]
        order = sorted(range(len(rc)), key=lambda i: (np.isnan(vals[i]), -vals[i] if not np.isnan(vals[i]) else 0))
        move = rc[order[:excess_bc]]
        roles[move] = "WV"

    is_bc = roles[rest] == "BC"
    need_bc = target - int(is_bc.sum())
    if need_bc > 0:
        rc = rest[roles[rest] == "WV"]
        vals = median_p[rc]
        order = sorted(range(len(rc)), key=lambda i: (np.isnan(vals[i]), vals[i] if not np.isnan(vals[i]) else 0))
        move = rc[order[:need_bc]]
        roles[move] = "BC"

    # ---- produce outputs: LABEL + SORTED Nodo ----
    nodo_labeled = [dict(s) for s in nodo]
    for k in range(n):
        nodo_labeled[k]["Label"] = str(roles[k])

    mbp_idx = np.flatnonzero(roles == "MBP")
    wv_idx = np.flatnonzero(roles == "WV")
    bc_idx = np.flatnonzero(roles == "BC")
    unk_idx = np.flatnonzero(roles == "Unknown")
    order = np.concatenate([mbp_idx, wv_idx, bc_idx, unk_idx]).astype(int)

    nodo_out = [nodo_labeled[i] for i in order]
    roles_table = [
        {
            "OriginalIdx": int(i),
            "ID": ids[i],
            "Label": str(roles[i]),
            "MedianP_Steady": float(median_p[i]),
        }
        for i in order
    ]

    print("\n[identify_brake_sensors]")
    print(f"  Steady window used: {tgrid[i1]} -> {tgrid[i2]} "
          f"({(tgrid[i2] - tgrid[i1]) / np.timedelta64(1, 'm'):.1f} min)")
    print(f"  Grid Fs = {fs_grid:.1f} Hz, LPF Fc = {fc:.1f} Hz, "
          f"|dP/dt| < {grad_tol:.3f} bar/s, MaxGap <= {max_gap_sec:.0f} s")
    print("  Classification by median pressure in steady window: >4.6=MBP, <0.2=BC, else=WV")
    for row in roles_table:
        print(f"    {row}")

    steady_info = SteadyInfo(
        candidate_found=True,
        tgrid=tgrid, p_filtered=Pf, dp_dt=dP, steady_all=steady_all,
        steady_mask=steady_mask, window_idx=(i1, i2),
        window_time=(tgrid[i1], tgrid[i2]), max_window_min=max_window_min,
    )
    return nodo_out, roles_table, steady_info
